src/array/cuda/spmat_op_impl_csr.cu

```diff
@@ -0,0 +1,538 @@
+  const IdType next = aten::IndexSelect<IdType>(csr.indptr, row + 1);
+  return next - cur;
+}
+
+template int64_t CSRGetRowNNZ<kDLGPU, int32_t>(CSRMatrix, int64_t);
+template int64_t CSRGetRowNNZ<kDLGPU, int64_t>(CSRMatrix, int64_t);
+
+template <typename IdType>
+__global__ void _CSRGetRowNNZKernel(
+    const IdType* vid,
+    const IdType* indptr,
+    IdType* out,
+    int64_t length) {
+  int tx = blockIdx.x * blockDim.x + threadIdx.x;
+  int stride_x = gridDim.x * blockDim.x;
+  while (tx < length) {
+    const IdType vv = vid[tx];
+    out[tx] = indptr[vv + 1] - indptr[vv];
+    tx += stride_x;
+  }
+}
+
+template <DLDeviceType XPU, typename IdType>
+NDArray CSRGetRowNNZ(CSRMatrix csr, NDArray rows) {
+  auto* thr_entry = runtime::CUDAThreadEntry::ThreadLocal();
+  const auto len = rows->shape[0];
+  const IdType* vid_data = static_cast<IdType*>(rows->data);
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  NDArray rst = NDArray::Empty({len}, rows->dtype, rows->ctx);
+  IdType* rst_data = static_cast<IdType*>(rst->data);
+  const int nt = cuda::FindNumThreads(len);
+  const int nb = (len + nt - 1) / nt;
+  _CSRGetRowNNZKernel<<<nb, nt, 0, thr_entry->stream>>>(
+      vid_data, indptr_data, rst_data, len);
+  return rst;
+}
+
+template NDArray CSRGetRowNNZ<kDLGPU, int32_t>(CSRMatrix, NDArray);
+template NDArray CSRGetRowNNZ<kDLGPU, int64_t>(CSRMatrix, NDArray);
+
+///////////////////////////// CSRGetRowColumnIndices /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+NDArray CSRGetRowColumnIndices(CSRMatrix csr, int64_t row) {
+  const int64_t len = impl::CSRGetRowNNZ<XPU, IdType>(csr, row);
+  const int64_t offset = aten::IndexSelect<IdType>(csr.indptr, row) * sizeof(IdType);
+  return csr.indices.CreateView({len}, csr.indices->dtype, offset);
+}
+
+template NDArray CSRGetRowColumnIndices<kDLGPU, int32_t>(CSRMatrix, int64_t);
+template NDArray CSRGetRowColumnIndices<kDLGPU, int64_t>(CSRMatrix, int64_t);
+
+///////////////////////////// CSRGetRowData /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+NDArray CSRGetRowData(CSRMatrix csr, int64_t row) {
+  const int64_t len = impl::CSRGetRowNNZ<XPU, IdType>(csr, row);
+  const int64_t offset = aten::IndexSelect<IdType>(csr.indptr, row) * sizeof(IdType);
+  if (aten::CSRHasData(csr))
+    return csr.data.CreateView({len}, csr.data->dtype, offset);
+  else
+    return aten::Range(offset, offset + len, csr.indptr->dtype.bits, csr.indptr->ctx);
+}
+
+template NDArray CSRGetRowData<kDLGPU, int32_t>(CSRMatrix, int64_t);
+template NDArray CSRGetRowData<kDLGPU, int64_t>(CSRMatrix, int64_t);
+
+///////////////////////////// CSRSliceRows /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+CSRMatrix CSRSliceRows(CSRMatrix csr, int64_t start, int64_t end) {
+  const int64_t num_rows = end - start;
+  const IdType st_pos = aten::IndexSelect<IdType>(csr.indptr, start);
```

Comment:
We can use CSRGetRowNNZ to compute nnz here.

Reply:
We only have the following CSRGetRowNNZ operators:
* `CSRGetRowNNZ(CSRMatrix, int64_t row)`
* `CSRGetRowNNZ(CSRMatrix, IdArray rows)`
, but we need `CSRGetRowNNZ(CSRMatrix, int64_t start, int64_t end)`. That's a reasonable ask but it can also be  implemented in two lines of codes quite easily.